File: Signal/src/phone/signaling/number directory/GroupContactsResult.m
//
//  GroupContactsResult.m
//  Signal
//
//  Created by Frederic Jacobs on 17/02/15.
//  Copyright (c) 2015 Open Whisper Systems. All rights reserved.
//

#import "GroupContactsResult.h"

#import <TextSecureKit/TSAccountManager.h>
#import "Contact.h"
#import "ContactsManager.h"
#import "Environment.h"
#import "SignalKeyingStorage.h"

@interface GroupContactsResult ()

@property NSMutableArray *unknownNumbers;
@property NSMutableArray *knownNumbers;

@property NSMutableDictionary *associatedContactDict;

@end

@implementation GroupContactsResult

- (instancetype)initWithMembersId:(NSArray *)memberIdentifiers without:(NSArray *)removeIds {
    self = [super init];

    ContactsManager *manager = [Environment.getCurrent contactsManager];

    NSMutableSet *remainingNumbers = [NSMutableSet setWithArray:memberIdentifiers];

    NSMutableArray *knownNumbers       = [NSMutableArray array];
    NSMutableArray *associatedContacts = [NSMutableArray array];

    for (NSString *identifier in memberIdentifiers) {
        if ([identifier isEqualToString:[TSAccountManager localNumber]]) {
            // remove local number

            [remainingNumbers removeObject:identifier];
            continue;
        }

        if (removeIds && [removeIds containsObject:identifier]) {
            // Remove ids
            [remainingNumbers removeObject:identifier];
            continue;
        }

        PhoneNumber *number = [PhoneNumber phoneNumberFromE164:identifier];

        if (!number) {
            continue;
        }

        Contact *contact = [manager latestContactForPhoneNumber:number];

        if (!contact) {
            continue;
        }

        [knownNumbers addObject:identifier];
        [associatedContacts addObject:contact];

        [remainingNumbers removeObject:identifier];
    }

    _unknownNumbers = [NSMutableArray arrayWithArray:[remainingNumbers allObjects]];
    [_unknownNumbers sortUsingComparator:^NSComparisonResult(id obj1, id obj2) {
      return [obj1 compare:obj2 options:0];
    }];


    // Populate mapping dictionary.
    _associatedContactDict = [NSMutableDictionary dictionary];
    for (NSUInteger i = 0; i < [knownNumbers count]; i++) {
        NSString *identifier = [knownNumbers objectAtIndex:i];
        Contact *contact     = [associatedContacts objectAtIndex:i];

        [_associatedContactDict setObject:contact forKey:identifier];
    }

    // Known Numbers
    _knownNumbers = [NSMutableArray arrayWithArray:knownNumbers];
    [_knownNumbers sortedArrayUsingComparator:^NSComparisonResult(id obj1, id obj2) {
      Contact *contact1 = [_associatedContactDict objectForKey:obj1];
      Contact *contact2 = [_associatedContactDict objectForKey:obj2];

      return [[manager class] contactComparator](contact1, contact2);
    }];

    return self;
}

- (NSUInteger)numberOfMembers {
    return [self.knownNumbers count] + [self.unknownNumbers count];
}

- (BOOL)isContactAtIndexPath:(NSIndexPath *)indexPath {
    if ((NSUInteger)indexPath.row < [self.unknownNumbers count]) {
        return NO;
    } else {
        return YES;
    }
}

- (Contact *)contactForIndexPath:(NSIndexPath *)indexPath {
    if ([self isContactAtIndexPath:indexPath]) {
        NSString *identifier = [_knownNumbers objectAtIndex:[self knownNumbersIndexForIndexPath:indexPath]];
        return [_associatedContactDict objectForKey:identifier];
    } else {
        NSAssert(NO, @"Trying to retrieve contact from array at an index that is not a contact.");
        return nil;
    }
}

- (NSString *)identifierForIndexPath:(NSIndexPath *)indexPath {
    if ([self isContactAtIndexPath:indexPath]) {
        return [_knownNumbers objectAtIndex:[self knownNumbersIndexForIndexPath:indexPath]];
    } else {
        return [_unknownNumbers objectAtIndex:(NSUInteger)indexPath.row];
    }
}

- (NSUInteger)knownNumbersIndexForIndexPath:(NSIndexPath *)indexPath {
    NSAssert(((NSUInteger)indexPath.row >= [_unknownNumbers count]), @"Wrong index for known number");

    return (NSUInteger)indexPath.row - [_unknownNumbers count];
}

@end

File: Signal/src/contact/GroupContactsResult.swift
//
//  GroupContactsResult.swift
//  Signal
//
//  Created by Tran Son on 4/10/17.
//  Copyright © 2017 Open Whisper Systems. All rights reserved.
//

import Foundation

@objc class GroupContactsResult: NSObject {

    private var unknownNumbers: [String]
    private var knownNumbers: [String]
    private var associatedContactDict: [String: Contact]

    init(withMembersId memberIdentifiers: [String], without removeIds: [String]) {
        let manager = Environment.getCurrent().contactsManager
        var remainingNumbers = Set(memberIdentifiers)
        var numbers = [String]()
        var associatedContacts = [Contact]()
        for identifier in memberIdentifiers {
            if identifier == TSAccountManager.localNumber() {
                // Remove local number
                remainingNumbers.remove(identifier)
                continue
            }
            if removeIds.contains(identifier) {
                // Remove id
                remainingNumbers.remove(identifier)
                continue
            }
            guard let number = PhoneNumber(fromE164: identifier), let contact = manager?.latestContact(for: number) else {
                continue
            }
            numbers.append(identifier)
            associatedContacts.append(contact)
            remainingNumbers.remove(identifier)
        }
        unknownNumbers = Array(remainingNumbers)
        unknownNumbers.sort(by: { return $0.compare($1) == .orderedAscending })
        // Populate mapping dictionary.
        var contactDict = [String: Contact]()
        for i in 0 ..< numbers.count {
            let identifier = numbers[i]
            let contact = associatedContacts[i]
            contactDict[identifier] = contact
        }
        // Known Numbers
        knownNumbers = Array(numbers)
        knownNumbers.sort(by: {
            (number1, number2) -> Bool in
            guard let contact1 = contactDict[number1], let contact2 = contactDict[number2] else {
                return false
            }
            return OWSContactsManager.contactComparator()(contact1, contact2) == .orderedAscending
        })
        associatedContactDict = contactDict
        super.init()
    }

    func numberOfMembers() -> Int {
        return knownNumbers.count + unknownNumbers.count
    }

    func isContact(at indexPath: IndexPath) -> Bool {
        return indexPath.row >= unknownNumbers.count
    }

    func contact(for indexPath: IndexPath) -> Contact? {
        if isContact(at: indexPath) {
            return associatedContactDict[knownNumbers[knownNumbersIndex(for: indexPath)]]
        } else {
            return nil
        }
    }

    func identifier(for indexPath: IndexPath) -> String? {
        return (isContact(at: indexPath) ? knownNumbers[knownNumbersIndex(for: indexPath)] : unknownNumbers[indexPath.row])
    }

    // MARK: - Helpers

    private func knownNumbersIndex(for indexPath: IndexPath) -> Int {
        guard indexPath.row >= unknownNumbers.count else {
            return 0
        }
        return indexPath.row - unknownNumbers.count
    }
}
